text/0000-additivescenes-and-networkscenemanager-refactoring.md

```diff
@@ -0,0 +1,247 @@
+        * [SynchronizeSceneNetworkObjects](https://github.com/Unity-Technologies/com.unity.multiplayer.mlapi/blob/bac7f416ae29624277984c6d2d1eee07bf4afb77/com.unity.multiplayer.mlapi/Runtime/SceneManagement/SceneEventData.cs#L518): A method of SceneEventData that is invoked from within the ClientLoadedSynchronization method.  This  handles the instantiation and synchronization of the associated scene’s NetworkObjects.
+ 3. Finally, a check is done to determine if there are still more scenes to be loaded.  If so, then the synchronization process repeats until the client has loaded all scenes and instantiated or soft-synchronized all NetworkObjects.  Once completed, the client responds to the server with a SceneEvent type of S2C_SyncComplete which includes the NetworkObjects instantiated and synchronized for the server to determine if a re-synchronization is required.
+
+### **The Entire Client Connection Approval and Scene Synchronization Process**
+We can see that the approval process does not include the scene synchronization data and will be processed as a completely separate message.  Then if EnableSceneManagement is true, the SynchronizeNetworkObjects method is invoked which generates the S2C_Sync SceneEvent message.  The client processes all serialized data (i.e. scenes and NetworkObjects) in the S2C_Sync message, and upon finishing informs the server that is is done (C2S_SyncComplete).
+![](0000-additivescenes-and-networkscenemanager-refactoring/NSM_Reference7.png)
+
+In the event that the server determines the client being synchronized missed one or more DestroyObject message(s), the server will send a final S2C_ReSync message that contains the NetworkObjectIds of the NetworkObjects that no longer exist.  Upon receiving the S2C_ReSync message, the client will remove the NetworkObjects in question and clean up the local SpawnManager's SpawnObjects lists.
+*This is a placed holder feature until the snapshot system handles the DestroyObject message*.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+### **The Client Scene Synchronization Process**
+With the above modifications, we can now consider the same problematic scenario with NetworkObjects that are dependent upon an additively loaded scene (diagram below).  While the server is creating the S2C_Sync SceneEvent, NetworkObjects are first grouped by their associated scene and then ordered such that NetworkObjects with dependencies will be sorted towards the end of the scene group specific serialized data.  This assures the spawn generators (i.e NetworkPrefabHandler)  will be instantiated before the dependent NetworkObjects.  
+
+![](0000-additivescenes-and-networkscenemanager-refactoring/NSM_Reference11.png)
+
+In order to set a scene dependency for a NetworkObject that has a dependency in a different scene (in this case the spawn generator), the user is only required to set the scene as a dependency for the dependent NetworkObject as such:
```

Comment:
here again, I'm not sure why we care about scene dependencies of networkobjects. if a developer wants to implement an object pool, put that into a separate scene, load that scene at runtime and spawn objects by that spawnpool to manage — why "netcode" needs to care about that at all? I'd say, that's something for implementor to figure out, if they have strong dependencies between objects being spawned/pooled and the spawnpool implementation. well, I think we could chat about that :)

Reply:
_(see the previous response above)_

Reply:
We no longer care about scene dependencies.  We now only care when certain NetworkObjects are soft synchronized or spawned, but that is only the special case of NetworkPrefabHandler related stuff.   In terms of scenes and dependencies that has been addressed in the most recent update in PR-955 and is reflected in this RFC.